packages/neo-one-cli/src/__data__/projects/ico/neo-one/contracts/Exchange.ts

```diff
@@ -0,0 +1,348 @@
+import {
+  Address,
+  constant,
+  Deploy,
+  MapStorage,
+  crypto,
+  Integer,
+  createEventNotifier,
+  SmartContract,
+} from '@neo-one/smart-contract';
+
+const notifyDeposited = createEventNotifier<Address, Address, Integer>('deposited', 'address', 'assetID', 'amount');
+const notifyWithdrawn = createEventNotifier<Address, Address, Integer>('withdrawn', 'address', 'assetID', 'amount');
+const notifyOfferCreated = createEventNotifier<Address, Address, Integer, Address, Integer, Buffer>(
+  'offerCreated',
+  'makerAddress',
+  'offerAssetID',
+  'offerAmount',
+  'wantAssetID',
+  'wantAmount',
+  'offerHash',
+);
+const notifyOfferCancelled = createEventNotifier<Buffer>('offerCancelled', 'offerHash');
+const notifyBurnt = createEventNotifier<Address, Address, Integer>(
+  'burnt',
+  'filler',
+  'takerFeeAssetID',
+  'takerFeeAmount',
+);
+const notifyFilled = createEventNotifier<Address, Buffer, Integer, Address, Integer, Address, Integer, Integer>(
+  'offerFilled',
+  'filler',
+  'offerHash',
+  'amountToFill',
+  'offerAssetID',
+  'offerAmount',
+  'wantAssetID',
+  'wantAmount',
+  'amountToTake',
+);
+
+interface NEP5 {
+  readonly transfer: (from: Address, to: Address, amount: Integer) => boolean;
+}
+
+type Offer = {
+  readonly maker: Address;
+  readonly offerAssetID: Address;
```

Comment:
This would be nice, but I think it makes it awkward to store.  Sub objects are not currently serializable objects, so I end up having to store the flattened version anyways.